### DISCUSSION TITLE
Should .transition() execute actions?

### DISCUSSION THREAD

**Discussion by kjartanm** [statelyai/xstate#2584] [Q&A]
Hi!
I'm doing a PoC with xstate on serverless, and so far I'm impressed with the possibilities!

But I have been struggling a bit with handling async-ness of invoked services in this scenario, so I thought i try the custom interpreter route. Doing that, it seems to me that I don't need to execute the actions manually after transitions? 

Because in all the testing I have done, the actions and following context changes happens anyway during `.transition(...)`. Is this as it should be, or am I missing something here?

For me it is not a big deal, but since the documentation says otherwise I thought I should ask.

**Comment by davidkpiano** [comment]
Excited that you're playing around with XState server-side! 🚀 

> ...the actions and following context changes happens anyway during .transition(...). Is this as it should be, or am I missing something here?

When you call `machine.transition(...)`, it is a pure function that returns the next state (including `context`) and all of the `actions` that _will_ be executed. It's up to you to execute them afterwards.

So **no, `.transition()` does not execute actions.**

This can be done like so:

```js
state.actions.forEach(action => {
  // execute the actions one by one
  // implementation of `executeAction` is up to you
  executeAction(action);
});
```

Does that answer your question?

**Reply by kjartanm** [reply]
Thanks for the quick reply :) !

What you describe is what I expected, but when testing it I discovered that I had commented out the execute part, but still got the context I expected. I tested again now with this smaller example in a REPL:

```
import { createMachine, assign } from 'xstate';

const countMachine = createMachine({
    id: 'counter',
    initial: 'count',
    context: {
        count: 0,
    },
    on: {
        RESET: {
            target: 'count',
            actions: ['reset'],
        },
    },
    states: {
        count: {
            entry: ['logEntry'],
            on: {
                ADD: {
                    target: 'count',
                    actions: ['add'],
                },
                SUBTRACT: {
                    target: 'count',
                    actions: ['subtract'],
                },
            },
        },
    },
},
{
    actions: {
        reset: assign({ count: (context, event) => 0}),
        add: assign({ count: (context, event) => context.count + 1}),
        subtract: assign({ count: (context, event) => context.count - 1}),
        logEntry: () => console.log("Enters state"),
    },
});

let newState = countMachine.transition('count', 'ADD');

const { actions } = newState;

for (let i = 0; i < actions.length; i++) {
    // If the action is executable, execute it
    const action = actions[i];
    if(typeof action.exec === 'function'){
       action.exec();
    }
}; 

console.log(newState.context); 
```
This will result in:
```
Enters state 

{ count: 1 }

```
If I comment out `//action.exec();`, I still get `{ count: 1 }` .

Is it only enter/exit-actions that get 'silenced'?

**Reply by davidkpiano** [reply]
Built-in "microstep" actions like `assign(...)` and `raise(...)` are evaluated as part of the transition. They're not exactly considered side-effects, since they determine what the next state (`state.context`) is going to be.

**Reply by kjartanm** [reply]
Ok, that makes sense. Thanks!
